ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java

```diff
@@ -63,4 +66,82 @@ private boolean isValid(AppendEntriesReplyProto reply) {
       return appender.getCallIdComparator().compare(reply.getServerReply().getCallId(), minCallId) >= 0;
     }
   }
+
+  static class HeartbeatBroadcastListener {
+    private final int groupPeerCount;
+    private int ack;
+    private int fail;
+    private volatile boolean done;
+    private final CompletableFuture<Boolean> result;
+
+    HeartbeatBroadcastListener(int groupSize) {
+      this.groupPeerCount = groupSize;
+      this.ack = 0;
+      this.fail = 0;
+      this.done = false;
+      result = new CompletableFuture<>();
+    }
+
+    private class HeartbeatListener implements BiFunction<Boolean, AppendEntriesReplyProto, Boolean> {
+      private final HeartbeatAck heartbeatAck;
+
+      HeartbeatListener(LogAppender logAppender) {
+        heartbeatAck = new HeartbeatAck(logAppender);
+      }
+
+      @Override
+      public Boolean apply(Boolean isTimeout, AppendEntriesReplyProto reply) {
+        if (isTimeout) {
+          onHeartbeatReply(false);
+          return true;
+        }
+
+        boolean acknowledged = heartbeatAck.receive(reply);
+        if (acknowledged) {
+          onHeartbeatReply(true);
+        }
+        return heartbeatAck.isAcknowledged();
+      }
+
+    }
+
+    HeartbeatListener newHeartbeatListener(LogAppender logAppender) {
+      return new HeartbeatListener(logAppender);
+    }
+
+    synchronized void onHeartbeatReply(boolean isSuccess) {
+      if (done) {
+        return;
+      }
+
+      if (isSuccess) {
+        ack++;
+      } else {
+        fail++;
+      }
+
+      if (isMajorityAck()) {
+        result.complete(true);
+        done = true;
+      }
+
+      if (isMajorityFail()) {
```

Comment:
There is not such thing as "majorityFail" since it should retry for failures.

Reply:
done